SystemMicrocontroller_Sensors/Core/Src/lcd20x4.c

```diff
@@ -0,0 +1,331 @@
+		GPIO_InitStruct.Pin = lcd->DB4_pin;
+		HAL_GPIO_Init(lcd->DB4_port, &GPIO_InitStruct);
+		GPIO_InitStruct.Pin = lcd->DB5_pin;
+		HAL_GPIO_Init(lcd->DB5_port, &GPIO_InitStruct);
+		GPIO_InitStruct.Pin = lcd->DB6_pin;
+		HAL_GPIO_Init(lcd->DB6_port, &GPIO_InitStruct);
+		GPIO_InitStruct.Pin = lcd->DB7_pin;
+		HAL_GPIO_Init(lcd->DB7_port, &GPIO_InitStruct);
+		HAL_GPIO_WritePin(lcd->DB4_port, lcd->DB4_pin, RESET);
+		HAL_GPIO_WritePin(lcd->DB5_port, lcd->DB5_pin, RESET);
+		HAL_GPIO_WritePin(lcd->DB6_port, lcd->DB6_pin, RESET);
+		HAL_GPIO_WritePin(lcd->DB7_port, lcd->DB7_pin, RESET);
+	} else {
+		GPIO_InitStruct.Pin = lcd->DB0_pin;
+		HAL_GPIO_Init(lcd->DB0_port, &GPIO_InitStruct);
+		GPIO_InitStruct.Pin = lcd->DB1_pin;
+		HAL_GPIO_Init(lcd->DB1_port, &GPIO_InitStruct);
+		GPIO_InitStruct.Pin = lcd->DB2_pin;
+		HAL_GPIO_Init(lcd->DB2_port, &GPIO_InitStruct);
+		GPIO_InitStruct.Pin = lcd->DB3_pin;
+		HAL_GPIO_Init(lcd->DB3_port, &GPIO_InitStruct);
+		GPIO_InitStruct.Pin = lcd->DB4_pin;
+		HAL_GPIO_Init(lcd->DB4_port, &GPIO_InitStruct);
+		GPIO_InitStruct.Pin = lcd->DB5_pin;
+		HAL_GPIO_Init(lcd->DB5_port, &GPIO_InitStruct);
+		GPIO_InitStruct.Pin = lcd->DB6_pin;
+		HAL_GPIO_Init(lcd->DB6_port, &GPIO_InitStruct);
+		GPIO_InitStruct.Pin = lcd->DB7_pin;
+		HAL_GPIO_Init(lcd->DB7_port, &GPIO_InitStruct);
+		HAL_GPIO_WritePin(lcd->DB0_port, lcd->DB0_pin, RESET);
+		HAL_GPIO_WritePin(lcd->DB1_port, lcd->DB1_pin, RESET);
+		HAL_GPIO_WritePin(lcd->DB2_port, lcd->DB2_pin, RESET);
+		HAL_GPIO_WritePin(lcd->DB3_port, lcd->DB3_pin, RESET);
+		HAL_GPIO_WritePin(lcd->DB4_port, lcd->DB4_pin, RESET);
+		HAL_GPIO_WritePin(lcd->DB5_port, lcd->DB5_pin, RESET);
+		HAL_GPIO_WritePin(lcd->DB6_port, lcd->DB6_pin, RESET);
+		HAL_GPIO_WritePin(lcd->DB7_port, lcd->DB7_pin, RESET);
+	}
+
+}
+
+
+void LCD_Enable_GPIO_RCC(LCD_HandleTypeDef * lcd) {
+	GPIO_TypeDef *CurrentGPIO;
+
+	CurrentGPIO = GPIOA;
+	if(isCurrentGPIOUsed){
+		__HAL_RCC_GPIOA_CLK_ENABLE();
+	}
+	CurrentGPIO = GPIOB;
+	if(isCurrentGPIOUsed){
+		__HAL_RCC_GPIOB_CLK_ENABLE();
+	}
+	CurrentGPIO = GPIOC;
+	if(isCurrentGPIOUsed){
+		__HAL_RCC_GPIOC_CLK_ENABLE();
+	}
+	CurrentGPIO = GPIOD;
+	if(isCurrentGPIOUsed){
+		__HAL_RCC_GPIOD_CLK_ENABLE();
+	}
+	CurrentGPIO = GPIOE;
+	if(isCurrentGPIOUsed){
+		__HAL_RCC_GPIOE_CLK_ENABLE();
+	}
+	CurrentGPIO = GPIOF;
+	if(isCurrentGPIOUsed){
+		__HAL_RCC_GPIOF_CLK_ENABLE();
+	}
+	CurrentGPIO = GPIOG;
+	if(isCurrentGPIOUsed){
+		__HAL_RCC_GPIOG_CLK_ENABLE();
+	}
+	CurrentGPIO = GPIOH;
+	if(isCurrentGPIOUsed){
+		__HAL_RCC_GPIOH_CLK_ENABLE();
+	}
+
+}
+
+//Prints to current cursor position
+void LCD_Printf(LCD_HandleTypeDef * lcd, const char* str, ...)
+{
+	char stringArray[80];
+	va_list args;
+	va_start(args, str);
+	vsprintf(stringArray, str, args);
+	va_end(args);
+	for(uint8_t i=0;  i<strlen(stringArray) && i<80; i++)
+	{
+		LCD_WriteData(lcd, (uint8_t)stringArray[i]);
+	}
+}
+
+void LCD_PrintTo1stLine(LCD_HandleTypeDef * lcd, const char* str, ...)
```

Comment:
Could the line be pass a parameter of the function or we need 4 different funtions?

Reply:
Definitely!